style!: issue15 - cleaner polynomial interface

samlaf opened the pull request:
Fixes #15 
Fixes #19 

Built on top of #18 so targetting that branch to make the PR only contain the code related to this issue.

Separate Polynomial into PolynomialEvalForm and PolynomialCoefForm to make library type safe when converting between these. Eg: fft can only go from coeff->eval, and ifft can only go from eval->coeff

This is just a proposal. There are still issues to address, as raised in #19 and #20. We'll meet on monday to discuss these. We might also want to change the srs code to use the same type-safe techniques (separate srs into monomial and lagrange form, such that fft/ifft only works on one of those).


Haven't fixed the tests/benchmarks yet, because waiting for review on style and our monday discussion. @anupsv can we change the settings of this repo to run ci on all branches. Guessing it's set to only run on PRs that target master right now?

bxue-l2 commented on src/kzg.rs:
```diff
@@ -484,62 +487,65 @@ impl Kzg {
-            },
-        };
+        // When the polynomial is in evaluation form, use the original g1 points
+        let bases = self.g1[..polynomial.len()].to_vec();
```
if it is eval form, then you should use lagrange basis. 

I suppose your g1 is monomial basis, downloaded from EigenDA

bxue-l2 replied:
see link, https://github.com/Layr-Labs/rust-kzg-bn254/issues/20#issuecomment-2570014488

samlaf replied:
Fixed, see https://github.com/Layr-Labs/rust-kzg-bn254/pull/21/files#diff-2f62ccf916a19a966ddd901f177604195dbb918aa16f4d7018e308d625e7f544R476

Made it so that we can only call the commit function now by using eval form polynomials (this was the implicit assumption in the previous code which wasnt tested anyways)

bxue-l2 commented on src/kzg.rs:
```diff
@@ -484,62 +487,65 @@ impl Kzg {
+        }
+        // When the polynomial is in coefficient form, use inverse FFT on the srs points.
+        // TODO: is this correct? See https://github.com/Layr-Labs/rust-kzg-bn254/issues/20
+        let bases = self.g1_ifft(polynomial.len())?;
```
same as above

samlaf replied:
Fixed, see https://github.com/Layr-Labs/rust-kzg-bn254/pull/21/files#diff-2f62ccf916a19a966ddd901f177604195dbb918aa16f4d7018e308d625e7f544R476

Made it so that we can only call the commit function now by using eval form polynomials (this was the implicit assumption in the previous code which wasnt tested anyways)

bxue-l2 commented on src/polynomial.rs:
```diff
@@ -1,137 +1,162 @@
+        helpers::to_byte_array(&self.evaluations, self.len_underlying_blob_bytes)
+    }
+
+    pub fn to_coef_form(&self) -> Result<PolynomialCoeffForm, PolynomialError> {
```
nix, use coeff, to be consistent

samlaf replied:
[c4c2e91](https://github.com/Layr-Labs/rust-kzg-bn254/pull/21/commits/c4c2e9134a76f686580e83616446b44eeab77c7a)

samlaf commented on README.md:
```diff
@@ -51,6 +47,15 @@ The `data_setup_custom` (for testing) or `data_setup_mins` should be used to spe
+![KZG Commitments](./kzg_commitment_diagram.png)
 
-The `compute_kzg_proof_with_roots_of_unity` takes in a `Polynomial` and an `index` at which it needs to be computed.
+Our current codebase has the types PolynomialEvalForm and PolynomialCoeffForm to represent the polynomial in evaluation and coefficient form respectively. However, we do not have types to represent the two forms of srs points. They are implicitly assumed to be in monomial form when loaded, and an IFFT is performed before taking the inner product with the polynomial in evaluation form.
```
Yes but I don't want to do it right now or in this PR. Created https://github.com/Layr-Labs/rust-kzg-bn254/issues/25

bxue-l2 replied:
don't think it is needed now

samlaf commented on src/polynomial.rs:
```diff
@@ -1,137 +1,162 @@
+    /// same length.
+    ///
+    /// TODO: We should get rid of this: polynomial should not know about the
+    /// blob.       This len is equivalent to the coeffs len before it gets
```
Not sure what you mean by safety. I think this is just a convenience function that @anupsv added to make it easy to get back the blob from the polynomial. Unclear where/how this is used though. We can remove it in some other PR.

bxue-l2 replied:
We have mapping from 
rollup data <-> blob
blob -> polynomial

because blob header contain the length of rollup data, I don't think we need the this length field
the kzg commits both blob header and codec(rollup_data) or everything is safe

bxue-l2 replied:
@anupsv what do you think

anupsv replied:
Yes, primarily a convenience method as this value is need to calculate the roots during the batch verification. As long as there is way to retrieve that information, i.e convert polynomial -> blob and then get this length, then no problem. This field would have avoided some repetitive work.

bxue-l2 replied:
if the data is padded already, you could just the root based on the vec length

samlaf replied:
@bxue-l2 you're missing a word there (just calculate the root?)

@anupsv why do we need to be able to revert polynomial -> blob? Shouldn't we always have the blob around wherever we convert to poly (which is typically just a quick conversion to be able to commit/generate proof afaiu?

An alternative proposal is to not pad the polynomial in the constructor, but only do it when needed (for eg when computing commit or proof). Only downside is if we need to do this a few times (which is probably always?).

anupsv replied:
@bxue-l2 yep we can, but it might mean adding new interface used by just 1 function. Just trying to keep the interfaces uniform. Or we can move everything to this interface.

@samlaf So in the deeper functions, per the spec, only polynomial is passed in. And since we won't have the data field in the poly, we'd need to get it from the blob, hence the reversion thought.

bxue-l2 replied:
Because the beginning of the blob contain the data, you can always figure out what the original data is, and get the length of codec blob length, by applying the codec function. It is a bit of detour, but you can do it

polynomial -> rollup-data -> codec blob

anupsv replied:
@bxue-l2 right thats also possible. But seems like a lot of work to get the length. If thats fine with everyone then we can work on that.

bxue-l2 replied:
we can open an issue on this problem, I don't think this field is removed. 

Look into the blob kind of breaking the library abstraction, which I don't like.

samlaf commented on src/polynomial.rs:
```diff
@@ -1,137 +1,162 @@
-    pub fn len(&self) -> usize {
-        self.elements.len()
+    /// TODO: we should deprecate this. See comment in the struct.
+    pub fn len_underlying_blob_bytes(&self) -> usize {
```
let's do this in another PR.

samlaf commented on src/kzg.rs:
```diff
@@ -462,55 +469,65 @@ impl Kzg {
-                self.g1_ifft(polynomial.len())?
-            },
-        };
+        // When the polynomial is in evaluation form, use IFFT to transform srs points
```
[88eb5ff](https://github.com/Layr-Labs/rust-kzg-bn254/pull/21/commits/88eb5ff37dc87ba7f8c61c041926d40396bde48a)

samlaf commented on src/kzg.rs:
```diff
@@ -284,8 +288,9 @@ impl Kzg {
-        // FIXME: Read the entire file (or large segments) into memory and then split it into field elements.
-        // Entire G1 file might be ~8GiB, so might not fit in RAM.
+        // FIXME: Read the entire file (or large segments) into memory and then split it
+        // into field elements. Entire G1 file might be ~8GiB, so might not fit
```
[464ac34](https://github.com/Layr-Labs/rust-kzg-bn254/pull/21/commits/464ac34c0f65d19928f819a43ae93b5d95a13559)

samlaf commented on src/blob.rs:
```diff
@@ -15,28 +14,31 @@ pub struct Blob {
-    /// to DA specs. If the data is already padded, use [`Blob::new`] instead.
+    /// Creates a new `Blob` from the provided raw_data byte slice and pads it
+    /// according to DA specs. If the data is already padded, use
+    /// [`Blob::new`] instead.
```
[09444ff](https://github.com/Layr-Labs/rust-kzg-bn254/pull/21/commits/09444ffc8e0bdd5362917e2cf4f9f32b7624fb05)

samlaf commented on README.md:
```diff
@@ -51,6 +47,15 @@ The `data_setup_custom` (for testing) or `data_setup_mins` should be used to spe
+## KZG Commitments
+
+Below diagram explains the difference types involved between polynomials, SRS points, and kzg commitments.
+A KZG commitment can be taken by an inner product between (poly_eval, srs_monomial) or (poly_coeff, srs_lagrange). FFT and IFFT operations can be performed to convert between these forms.
```
done

bxue-l2 commented:
looks good to me, want others to take a pass

anupsv commented on src/polynomial.rs:
```diff
@@ -1,137 +1,162 @@
+        let mut padded_evals = evals;
+        // TODO: does it make sense to pad evaluations with zeros? This changes the
+        // polynomial...
+        padded_evals.resize(next_power_of_two, Fr::zero());
```
If im understanding this right, when we call `to_eval_form()`, the Fr field size is changed to the next power of 2 and then padded with the "0" field element right ?

samlaf replied:
Not the Fr field size, but the size of the polynomial (number of evaluation points) is extended to the next power of 2 usings 0s yes. See the comment [here](https://github.com/Layr-Labs/rust-kzg-bn254/blob/style--issue15-cleaner-polynomial-interface/src/polynomial.rs#L86-89)

anupsv replied:
So if i have [1, 2, 3] in coeff, it gets padded to [1,2,3,0] and when `to_eval()` is called, does it do [1, 2, 3, 0, 0, 0, 0, 0] ?

samlaf replied:
correct because we use https://doc.rust-lang.org/std/primitive.usize.html#method.next_power_of_two
```
Returns the smallest power of two greater than or equal to self.
```

samlaf replied:
Added test: [e61ddcf](https://github.com/Layr-Labs/rust-kzg-bn254/pull/21/commits/e61ddcf7a1a89176e1c980a907ead0f26a33a6c8)

anupsv replied:
I see. I'm thinking we shouldn't pad extra 0's when calling `to_eval()` or `to_coeff()` right ? The fft or ifft will return the same amount of elements and we should be using only those after conversion ?
Or is it the understanding that since it's 0's the commitment will remain the same and the caller will only parse the number of bytes as specified by the header byte, which contains the length, attached by the proxy ?

bxue-l2 replied:
yes, that is what we are doing in hokelea, we get the entire blob and just read the data based on the header that encodes how much is the user data
https://github.com/Layr-Labs/hokulea/blob/master/crates/eigenda/src/eigenda_data.rs#L41

anupsv replied:
Ah I see. Then let's just add a comment in the code to capture this behaviour and be done ?

samlaf replied:
> I see. I'm thinking we shouldn't pad extra 0's when calling to_eval() or to_coeff() right ? The fft or ifft will return the same amount of elements and we should be using only those after conversion ?

it won't pad extra 0s, because the evals/coeffs are already a power of 2, so that won't change.

> Or is it the understanding that since it's 0's the commitment will remain the same and the caller will only parse the number of bytes as specified by the header byte, which contains the length, attached by the proxy ?

not sure I understand what you mean here

samlaf commented on src/blob.rs:
```diff
@@ -1,67 +1,49 @@
+    /// WARNING: This function does not check if the bytes are modulo bn254
+    /// if the data has 32 byte segments exceeding the modulo of the field
+    /// then the bytes will be modded by the order of the field and the data
+    /// will be transformed incorrectly.
```
its not about the length. its about if you interpret a 32 byte segment as a big endian number, we just assume those are meant to be bn254 elements, so they will be modded by the prime of that field.

I don't think we should be doing this, hence the TODO below. But think we should make further changes in a future PRl

samlaf replied:
will resolve this as there are too many parallel conversations on this PR. Let me know if that didn't make sense.

bxue-l2 reviewed (commented):
lgtm

bxue-l2 commented on README.md:
```diff
@@ -32,30 +29,39 @@ See the `test_compute_kzg_proof` function in [./tests/kzg_test.rs](./tests/kzg_t
+  - meaningful when committing and generating/verifying proofs
+  - Obtained from Blob by appending zeroes to make the length a power of 2, and then interpreting the array as coefficients or evaluations of a polynomial.
+- Kzg: struct storing the SRS points used to generate commitments and proofs
+- SRS points: bn254 group elements
```
ok, you noted below

bxue-l2 commented on README.md:
```diff
@@ -32,25 +29,39 @@ See the `test_compute_kzg_proof` function in [./tests/kzg_test.rs](./tests/kzg_t
+  - inner producted with the polynomial to generate commitments
 
-From the `Blob`, a polynomial can be obtained via calling the `to_polynomial()` function. This converts the Blob to Field elements, then calculates the next power of 2 from this length of field elements and appends `zero` value elements for the remaining length.
+The [Blob](./src/blob.rs) and [Polynomial](./src/polynomial.rs) structs are mostly [Plain Old Data](https://en.wikipedia.org/wiki/Passive_data_structure) with constructor and few helper methods. The interesting stuff happens in the [KZG](./src/kzg.rs) struct, which has methods for committing to a blob, polynomial in coeff or eval form, and generating and verifying proofs.
```
make sure to note that this blob is different from the blob defintion from the eigenda_client. I would get super confused by it
https://github.com/Layr-Labs/eigenda/blob/master/api/clients/eigenda_client.go#L167

samlaf replied:
I think this is too deep in the weeds. Nobody will know about this (for now). Let's just fix that name and not mention it here. Let's try to keep the docs here self-contained.

bxue-l2 replied:
Just say not to confuse with the others. Then once the interface update, we ca update here?

samlaf replied:
But we will just forget to update here and then confuse people even more. Let's not confuse them with problems they don't have haha.

samlaf commented:
Ok sorry guys I'm being fookin stupid keep pushing to a PR. Promise this was my last one. Decided to move all the new docs to an actual rust docs in lib.rs so that it'll show up on https://docs.rs/rust-kzg-bn254/latest/rust_kzg_bn254/.

See [194b32b](https://github.com/Layr-Labs/rust-kzg-bn254/pull/21/commits/194b32b849d9a097e75ba97e0c97dbd0a0d5fc81)

Let's merge this, then publish a new crate version to update the docs, and then we can make more small improvements in future PRs. Promise to not touch this PR anymore other than for addressing comments!
